{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "289fe589-7c9f-4412-bf34-ee912b9d1c8f",
   "metadata": {},
   "source": [
    "# Mapping data to a subset of the Whole Mouse Brain taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30bd55ac",
   "metadata": {},
   "source": [
    "Navigate to this link: https://alleninstitute.github.io/abc_atlas_access/intro.html and decide on the reference you wish to use. \n",
    "\n",
    "For this tutorial, we will be using the \"Mouse whole-brain transcriptomic cell type atlas (Hongkui Zeng)\" and specifically the 10xv3 dataset: https://alleninstitute.github.io/abc_atlas_access/descriptions/WMB-10Xv3.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57196e34",
   "metadata": {},
   "source": [
    "## A. Downloading the data\n",
    "### You need internet access to run the chunks below\n",
    "You need to install: \n",
    "- Expression matrices (.h5ad files containing raw data) - in our example there will be one object per brain region\n",
    "- Cell metadata files \n",
    "- Taxonomy files\n",
    "\n",
    "The simplest way to do this is to navigate to the corresponding page for each of those file categories in the AWS S3 explorer: https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/index.html \n",
    "\n",
    "For this tutorial, the corresponding links are: \n",
    "- Expression matrices: https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/index.html#expression_matrices/WMB-10Xv3/20230630/\n",
    "- Cell metadata (I'm using the most recent file): https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/metadata/WMB-10X/20241115/cell_metadata.csv\n",
    "- Taxonomy files: https://allen-brain-cell-atlas.s3.us-west-2.amazonaws.com/index.html#metadata/WMB-taxonomy/20231215/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3335e018",
   "metadata": {},
   "source": [
    "We will also need some additional files: \n",
    "- precomputed stats \n",
    "- marker lookup table\n",
    "\n",
    "You can get these by running the code below (Whole Mouse Brain taxonomy):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "482c0b65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-10-12 14:57:13--  https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/mapmycells/WMB-10X/20240831/mouse_markers_230821.json\n",
      "Resolving allen-brain-cell-atlas.s3-us-west-2.amazonaws.com... 52.218.152.49, 3.5.86.223, 52.92.160.10, ...\n",
      "Connecting to allen-brain-cell-atlas.s3-us-west-2.amazonaws.com|52.218.152.49|:443... ^C\n",
      "--2025-10-12 14:57:16--  https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/mapmycells/WMB-10X/20240831/precomputed_stats_ABC_revision_230821.h5\n",
      "Resolving allen-brain-cell-atlas.s3-us-west-2.amazonaws.com... 52.92.177.210, 3.5.76.167, 3.5.77.48, ...\n",
      "Connecting to allen-brain-cell-atlas.s3-us-west-2.amazonaws.com|52.92.177.210|:443... ^C\n"
     ]
    }
   ],
   "source": [
    "!wget https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/mapmycells/WMB-10X/20240831/mouse_markers_230821.json \n",
    "!wget https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/mapmycells/WMB-10X/20240831/precomputed_stats_ABC_revision_230821.h5 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdb5e40e",
   "metadata": {},
   "source": [
    "Here's the structure of my data folder:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fb253c25",
   "metadata": {
    "tags": [
     "parameters"
    ],
    "vscode": {
     "languageId": "bat"
    }
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid character '├' (U+251C) (1239423222.py, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[3], line 2\u001b[0;36m\u001b[0m\n\u001b[0;31m    ├── data\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid character '├' (U+251C)\n"
     ]
    }
   ],
   "source": [
    "ABC_atlas_celltyping/\n",
    "├── data\n",
    "│   ├── expression_matrices\n",
    "│   │   └── WMB-10Xv3\n",
    "│   │       └── 20230630\n",
    "│   │           ├── WMB-10Xv3-CTXsp-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-HPF-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-HY-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-Isocortex-1-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-Isocortex-2-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-MB-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-OLF-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-PAL-raw.h5ad\n",
    "│   │           ├── WMB-10Xv3-STR-raw.h5ad\n",
    "│   │           └── WMB-10Xv3-TH-raw.h5ad\n",
    "│   └── metadata\n",
    "│       ├── WMB-10X\n",
    "│       │   ├── 20230630\n",
    "│       │   │   └── cell_metadata.csv\n",
    "│       │   └── 20241115\n",
    "│       │       └── cell_metadata.csv\n",
    "│       └── WMB-taxonomy\n",
    "│           └── 20231215\n",
    "│               ├── cluster_annotation_term.csv\n",
    "│               ├── cluster_annotation_term_set.csv\n",
    "│               ├── cluster.csv\n",
    "│               ├── cluster_to_cluster_annotation_membership.csv\n",
    "│               └── views\n",
    "│                   ├── cluster_annotation_term_with_counts.csv\n",
    "│                   ├── cluster_to_cluster_annotation_membership_color.csv\n",
    "│                   └── cluster_to_cluster_annotation_membership_pivoted.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6628e74",
   "metadata": {},
   "source": [
    "## B. Subsetting the data to the desired regions\n",
    "### You can connect the kernel to an (offline) compute node for a faster analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0334ddb",
   "metadata": {},
   "source": [
    "Imports and folder organization:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c7dd19ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import anndata\n",
    "import pathlib\n",
    "import os\n",
    "import numpy as np\n",
    "import re\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "578481c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "abc_atlas_dir = pathlib.Path('/scratch/mfafouti/ABC_atlas_celltyping')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "30065c67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the path where the repository was cloned: /scratch/mfafouti/ABC_atlas_celltyping It will be used as a root for all subsequent paths\n"
     ]
    }
   ],
   "source": [
    "project_path = str(abc_atlas_dir)\n",
    "print(f'This is the path where the repository was cloned: {project_path} It will be used as a root for all subsequent paths')\n",
    "\n",
    "data_dir = abc_atlas_dir / 'data'\n",
    "expression_dir = data_dir / 'expression_matrices' / 'WMB-10Xv3' / '20230630'\n",
    "metadata_dir = data_dir / 'metadata' / 'WMB-10X' / '20241115'\n",
    "taxonomy_dir = data_dir / 'metadata' / 'WMB-taxonomy' / '20231215'\n",
    "scratch_dir = abc_atlas_dir / 'scratch'\n",
    "precompute_dir = scratch_dir / 'precompute'\n",
    "reference_dir = scratch_dir / 'reference'\n",
    "query_dir = data_dir / 'query_data'\n",
    "\n",
    "for dir_name in (scratch_dir, precompute_dir, reference_dir, query_dir):\n",
    "    if not dir_name.exists():\n",
    "        dir_name.mkdir()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ea1d4308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/query_data/BC13_collapsed_mouse_genes.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/query_data/BC14_collapsed_mouse_genes.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/query_data/BC15_collapsed_mouse_genes.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/query_data/BC28_collapsed_mouse_genes.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/query_data/BC3_collapsed_mouse_genes.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/query_data/BC9_collapsed_mouse_genes.h5ad')]\n"
     ]
    }
   ],
   "source": [
    "# query_file = data_dir / 'query_data'/ 'BC3_collapsed_mouse_genes.h5ad'\n",
    "query_file_list = list(query_dir.glob(\"*.h5ad\"))\n",
    "print(query_file_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9f8ca30",
   "metadata": {},
   "source": [
    "Below I define the regions I wish to include - I have excluded: CB (=cerebellum), MY (=medulla), P (=pons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6e139867",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_to_include = [\"CTXsp\",\"HPF\",\"HY\",\"Isocortex-1\",\"Isocortex-2\",\"MB\",\"OLF\",\"PAL\",\"STR\",\"TH\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4b6bb4c-1896-4e04-bbd5-d9337f622280",
   "metadata": {},
   "source": [
    "First we will preview the metadata CSV:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "29611974",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_label</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_cell_sample_label</th>\n",
       "      <th>library_label</th>\n",
       "      <th>feature_matrix_label</th>\n",
       "      <th>entity</th>\n",
       "      <th>brain_section_label</th>\n",
       "      <th>library_method</th>\n",
       "      <th>region_of_interest_acronym</th>\n",
       "      <th>donor_label</th>\n",
       "      <th>donor_genotype</th>\n",
       "      <th>donor_sex</th>\n",
       "      <th>dataset_label</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>cluster_alias</th>\n",
       "      <th>abc_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCGAGAAGTTAAGGGC-410_B05</td>\n",
       "      <td>GCGAGAAGTTAAGGGC</td>\n",
       "      <td>410_B05</td>\n",
       "      <td>L8TX_201030_01_C12</td>\n",
       "      <td>WMB-10Xv3-HPF</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>RHP</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-550850</td>\n",
       "      <td>Ai14(RCL-tdT)/wt</td>\n",
       "      <td>F</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>23.146826</td>\n",
       "      <td>-3.086639</td>\n",
       "      <td>1</td>\n",
       "      <td>484be5df-5d44-4bfe-9652-7b5bc739c211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATGGCTCAGCTCCTT-411_B06</td>\n",
       "      <td>AATGGCTCAGCTCCTT</td>\n",
       "      <td>411_B06</td>\n",
       "      <td>L8TX_201029_01_E10</td>\n",
       "      <td>WMB-10Xv3-HPF</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>RHP</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-550851</td>\n",
       "      <td>Ai14(RCL-tdT)/wt</td>\n",
       "      <td>F</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>23.138481</td>\n",
       "      <td>-3.022000</td>\n",
       "      <td>1</td>\n",
       "      <td>5638505d-e1e8-457f-9e5b-59e3e2302417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AACACACGTTGCTTGA-410_B05</td>\n",
       "      <td>AACACACGTTGCTTGA</td>\n",
       "      <td>410_B05</td>\n",
       "      <td>L8TX_201030_01_C12</td>\n",
       "      <td>WMB-10Xv3-HPF</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>RHP</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-550850</td>\n",
       "      <td>Ai14(RCL-tdT)/wt</td>\n",
       "      <td>F</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>23.472557</td>\n",
       "      <td>-2.992709</td>\n",
       "      <td>1</td>\n",
       "      <td>a0544e29-194f-4d34-9af4-13e7377b648f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CACAGATAGAGGCGGA-410_A05</td>\n",
       "      <td>CACAGATAGAGGCGGA</td>\n",
       "      <td>410_A05</td>\n",
       "      <td>L8TX_201029_01_A10</td>\n",
       "      <td>WMB-10Xv3-HPF</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>RHP</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-550850</td>\n",
       "      <td>Ai14(RCL-tdT)/wt</td>\n",
       "      <td>F</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>23.379622</td>\n",
       "      <td>-3.043442</td>\n",
       "      <td>1</td>\n",
       "      <td>c777ac0b-77e1-4d76-bf8e-2b3d9e08b253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAGTGAAGCATTTCG-410_B05</td>\n",
       "      <td>AAAGTGAAGCATTTCG</td>\n",
       "      <td>410_B05</td>\n",
       "      <td>L8TX_201030_01_C12</td>\n",
       "      <td>WMB-10Xv3-HPF</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv3</td>\n",
       "      <td>RHP</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-550850</td>\n",
       "      <td>Ai14(RCL-tdT)/wt</td>\n",
       "      <td>F</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>23.909480</td>\n",
       "      <td>-2.601536</td>\n",
       "      <td>1</td>\n",
       "      <td>49860925-e82b-46df-a228-fd2f97e75d39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cell_label      cell_barcode barcoded_cell_sample_label  \\\n",
       "0  GCGAGAAGTTAAGGGC-410_B05  GCGAGAAGTTAAGGGC                    410_B05   \n",
       "1  AATGGCTCAGCTCCTT-411_B06  AATGGCTCAGCTCCTT                    411_B06   \n",
       "2  AACACACGTTGCTTGA-410_B05  AACACACGTTGCTTGA                    410_B05   \n",
       "3  CACAGATAGAGGCGGA-410_A05  CACAGATAGAGGCGGA                    410_A05   \n",
       "4  AAAGTGAAGCATTTCG-410_B05  AAAGTGAAGCATTTCG                    410_B05   \n",
       "\n",
       "        library_label feature_matrix_label entity  brain_section_label  \\\n",
       "0  L8TX_201030_01_C12        WMB-10Xv3-HPF   cell                  NaN   \n",
       "1  L8TX_201029_01_E10        WMB-10Xv3-HPF   cell                  NaN   \n",
       "2  L8TX_201030_01_C12        WMB-10Xv3-HPF   cell                  NaN   \n",
       "3  L8TX_201029_01_A10        WMB-10Xv3-HPF   cell                  NaN   \n",
       "4  L8TX_201030_01_C12        WMB-10Xv3-HPF   cell                  NaN   \n",
       "\n",
       "  library_method region_of_interest_acronym                   donor_label  \\\n",
       "0          10Xv3                        RHP  Snap25-IRES2-Cre;Ai14-550850   \n",
       "1          10Xv3                        RHP  Snap25-IRES2-Cre;Ai14-550851   \n",
       "2          10Xv3                        RHP  Snap25-IRES2-Cre;Ai14-550850   \n",
       "3          10Xv3                        RHP  Snap25-IRES2-Cre;Ai14-550850   \n",
       "4          10Xv3                        RHP  Snap25-IRES2-Cre;Ai14-550850   \n",
       "\n",
       "     donor_genotype donor_sex dataset_label          x         y  \\\n",
       "0  Ai14(RCL-tdT)/wt         F     WMB-10Xv3  23.146826 -3.086639   \n",
       "1  Ai14(RCL-tdT)/wt         F     WMB-10Xv3  23.138481 -3.022000   \n",
       "2  Ai14(RCL-tdT)/wt         F     WMB-10Xv3  23.472557 -2.992709   \n",
       "3  Ai14(RCL-tdT)/wt         F     WMB-10Xv3  23.379622 -3.043442   \n",
       "4  Ai14(RCL-tdT)/wt         F     WMB-10Xv3  23.909480 -2.601536   \n",
       "\n",
       "   cluster_alias                         abc_sample_id  \n",
       "0              1  484be5df-5d44-4bfe-9652-7b5bc739c211  \n",
       "1              1  5638505d-e1e8-457f-9e5b-59e3e2302417  \n",
       "2              1  a0544e29-194f-4d34-9af4-13e7377b648f  \n",
       "3              1  c777ac0b-77e1-4d76-bf8e-2b3d9e08b253  \n",
       "4              1  49860925-e82b-46df-a228-fd2f97e75d39  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_metadata = pd.read_csv(metadata_dir / 'cell_metadata_new.csv')\n",
    "cell_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a4f342e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4042976, 17)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "219f2a42",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4071788/3966033639.py:5: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  cell_metadata.feature_matrix_label.str.contains(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered cells shape: (1824724, 17)\n",
      "Unique feature_matrix_labels after filtering:\n",
      "['WMB-10Xv3-HPF' 'WMB-10Xv3-Isocortex-1' 'WMB-10Xv3-PAL' 'WMB-10Xv3-STR'\n",
      " 'WMB-10Xv3-CTXsp' 'WMB-10Xv3-HY' 'WMB-10Xv3-OLF' 'WMB-10Xv3-TH'\n",
      " 'WMB-10Xv3-MB' 'WMB-10Xv3-Isocortex-2']\n"
     ]
    }
   ],
   "source": [
    "pattern = \"(\" + \"|\".join(re.escape(r) for r in regions_to_include) + \")\"\n",
    "\n",
    "# Match only WMB-10Xv3- datasets that contain those region names\n",
    "filtered_cells = cell_metadata[\n",
    "    cell_metadata.feature_matrix_label.str.contains(\n",
    "        f\"WMB-10Xv3-.*{pattern}\", case=False, na=False\n",
    "    )\n",
    "]\n",
    "print(f\"Filtered cells shape: {filtered_cells.shape}\")\n",
    "print(\"Unique feature_matrix_labels after filtering:\")\n",
    "print(filtered_cells.feature_matrix_label.unique()) # make sure this matches the regions of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7cae051",
   "metadata": {},
   "source": [
    "Load the expression matrices for the regions of interest:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7ef55a08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-CTXsp-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-HPF-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-HY-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-Isocortex-1-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-Isocortex-2-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-MB-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-OLF-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-PAL-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-STR-raw.h5ad'), PosixPath('/scratch/mfafouti/ABC_atlas_celltyping/data/expression_matrices/WMB-10Xv3/20230630/WMB-10Xv3-TH-raw.h5ad')]\n"
     ]
    }
   ],
   "source": [
    "h5ad_path_list = list(expression_dir.glob(\"*.h5ad\"))\n",
    "print(h5ad_path_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7464169a",
   "metadata": {},
   "source": [
    "## D. Create the files necessary to run cell_type_mapper\n",
    "The cell_type_mapper package: https://github.com/AllenInstitute/cell_type_mapper/tree/main requires the following input: \n",
    "- precomputed stats file \n",
    "- marker lookup files \n",
    "\n",
    "The files already exist for the whole taxonomy. There are 3 options\n",
    "1. Map to the full taxonomy \n",
    "(fastest, see https://github.com/AllenInstitute/cell_type_mapper/blob/67b740991ad30e2369f14aeef6f711e195a0ae76/examples/map_to_subset_of_taxonomy_fast.ipynb#L8)\n",
    "2. Create a custom version of those files \n",
    "(more complicated, not in this tutorial, see: https://github.com/AllenInstitute/cell_type_mapper/blob/67b740991ad30e2369f14aeef6f711e195a0ae76/examples/mapping_to_subset_of_abc_atlas_data.ipynb) \n",
    "3. Instruct the cell type mapper at mapping to ignore certain branches of the cell type taxonomy. \n",
    "\n",
    "We will use approach 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "dee8f34d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the dataframe that associates cells with cell type taxons\n",
    "taxonomy_path = taxonomy_dir / 'cluster_to_cluster_annotation_membership.csv'\n",
    "taxonomy_df = pd.read_csv(taxonomy_path)\n",
    "\n",
    "# create a dict mapping cluster_alias to taxons in the cell\n",
    "# type taxonomy\n",
    "alias_to_truth = dict()\n",
    "for cell in taxonomy_df.to_dict(orient='records'):\n",
    "    alias = cell['cluster_alias']\n",
    "    level = cell['cluster_annotation_term_set_label']\n",
    "    node = cell['cluster_annotation_term_label']\n",
    "    if alias not in alias_to_truth:\n",
    "        alias_to_truth[alias] = dict()\n",
    "    alias_to_truth[alias][level] = node\n",
    "\n",
    "# use the association between cells and cluster_alias to create a dict\n",
    "# mapping cell_label to cell type taxons\n",
    "# ground_truth = {\n",
    "#     cell_label: alias_to_truth[cluster_alias]\n",
    "#     for cell_label, cluster_alias in\n",
    "#     zip(test_set.cell_label.values, test_set.cluster_alias.values)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "21a9b0c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=======example nodes being dropped=======\n",
      "('subclass', 'CS20230722_SUBC_299')\n",
      "('subclass', 'CS20230722_SUBC_310')\n",
      "('subclass', 'CS20230722_SUBC_244')\n",
      "('subclass', 'CS20230722_SUBC_296')\n",
      "('subclass', 'CS20230722_SUBC_305')\n",
      "('subclass', 'CS20230722_SUBC_294')\n",
      "('subclass', 'CS20230722_SUBC_236')\n",
      "('subclass', 'CS20230722_SUBC_239')\n",
      "('subclass', 'CS20230722_SUBC_283')\n",
      "('subclass', 'CS20230722_SUBC_257')\n",
      "('subclass', 'CS20230722_SUBC_266')\n",
      "('subclass', 'CS20230722_SUBC_278')\n",
      "('subclass', 'CS20230722_SUBC_258')\n",
      "('subclass', 'CS20230722_SUBC_288')\n",
      "('subclass', 'CS20230722_SUBC_284')\n",
      "('subclass', 'CS20230722_SUBC_233')\n",
      "('subclass', 'CS20230722_SUBC_247')\n",
      "('subclass', 'CS20230722_SUBC_242')\n",
      "('subclass', 'CS20230722_SUBC_248')\n",
      "('subclass', 'CS20230722_SUBC_253')\n",
      "('subclass', 'CS20230722_SUBC_260')\n",
      "('subclass', 'CS20230722_SUBC_292')\n",
      "('subclass', 'CS20230722_SUBC_291')\n",
      "('subclass', 'CS20230722_SUBC_297')\n",
      "('subclass', 'CS20230722_SUBC_300')\n",
      "('subclass', 'CS20230722_SUBC_302')\n",
      "('subclass', 'CS20230722_SUBC_218')\n",
      "('subclass', 'CS20230722_SUBC_286')\n",
      "('subclass', 'CS20230722_SUBC_301')\n",
      "('subclass', 'CS20230722_SUBC_240')\n",
      "('subclass', 'CS20230722_SUBC_274')\n",
      "('subclass', 'CS20230722_SUBC_306')\n",
      "('subclass', 'CS20230722_SUBC_227')\n",
      "('subclass', 'CS20230722_SUBC_282')\n",
      "('subclass', 'CS20230722_SUBC_303')\n",
      "('subclass', 'CS20230722_SUBC_298')\n",
      "('subclass', 'CS20230722_SUBC_157')\n",
      "('subclass', 'CS20230722_SUBC_279')\n",
      "('subclass', 'CS20230722_SUBC_280')\n",
      "('subclass', 'CS20230722_SUBC_255')\n",
      "('subclass', 'CS20230722_SUBC_254')\n",
      "('subclass', 'CS20230722_SUBC_304')\n",
      "('subclass', 'CS20230722_SUBC_249')\n",
      "('subclass', 'CS20230722_SUBC_238')\n",
      "('subclass', 'CS20230722_SUBC_241')\n",
      "('subclass', 'CS20230722_SUBC_307')\n",
      "('subclass', 'CS20230722_SUBC_259')\n",
      "('subclass', 'CS20230722_SUBC_256')\n"
     ]
    }
   ],
   "source": [
    "valid_classes = set(\n",
    "    [\n",
    "        alias_to_truth[cl]['CCN20230722_SUBC']\n",
    "        for cl in filtered_cells.cluster_alias.values\n",
    "    ]\n",
    ")\n",
    "classes_to_drop = list(\n",
    "    set(\n",
    "        [alias_to_truth[cl]['CCN20230722_SUBC']\n",
    "         for cl in alias_to_truth\n",
    "         if alias_to_truth[cl]['CCN20230722_SUBC'] not in valid_classes]\n",
    "    )\n",
    ")\n",
    "\n",
    "nodes_to_drop = [('subclass', cl) for cl in classes_to_drop]\n",
    "\n",
    "print('=======example nodes being dropped=======')\n",
    "for pair in nodes_to_drop:\n",
    "    print(pair)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "455ed440",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('subclass', 'CS20230722_SUBC_299'), ('subclass', 'CS20230722_SUBC_310'), ('subclass', 'CS20230722_SUBC_244'), ('subclass', 'CS20230722_SUBC_296'), ('subclass', 'CS20230722_SUBC_305'), ('subclass', 'CS20230722_SUBC_294'), ('subclass', 'CS20230722_SUBC_236'), ('subclass', 'CS20230722_SUBC_239'), ('subclass', 'CS20230722_SUBC_283'), ('subclass', 'CS20230722_SUBC_257'), ('subclass', 'CS20230722_SUBC_266'), ('subclass', 'CS20230722_SUBC_278'), ('subclass', 'CS20230722_SUBC_258'), ('subclass', 'CS20230722_SUBC_288'), ('subclass', 'CS20230722_SUBC_284'), ('subclass', 'CS20230722_SUBC_233'), ('subclass', 'CS20230722_SUBC_247'), ('subclass', 'CS20230722_SUBC_242'), ('subclass', 'CS20230722_SUBC_248'), ('subclass', 'CS20230722_SUBC_253'), ('subclass', 'CS20230722_SUBC_260'), ('subclass', 'CS20230722_SUBC_292'), ('subclass', 'CS20230722_SUBC_291'), ('subclass', 'CS20230722_SUBC_297'), ('subclass', 'CS20230722_SUBC_300'), ('subclass', 'CS20230722_SUBC_302'), ('subclass', 'CS20230722_SUBC_218'), ('subclass', 'CS20230722_SUBC_286'), ('subclass', 'CS20230722_SUBC_301'), ('subclass', 'CS20230722_SUBC_240'), ('subclass', 'CS20230722_SUBC_274'), ('subclass', 'CS20230722_SUBC_306'), ('subclass', 'CS20230722_SUBC_227'), ('subclass', 'CS20230722_SUBC_282'), ('subclass', 'CS20230722_SUBC_303'), ('subclass', 'CS20230722_SUBC_298'), ('subclass', 'CS20230722_SUBC_157'), ('subclass', 'CS20230722_SUBC_279'), ('subclass', 'CS20230722_SUBC_280'), ('subclass', 'CS20230722_SUBC_255'), ('subclass', 'CS20230722_SUBC_254'), ('subclass', 'CS20230722_SUBC_304'), ('subclass', 'CS20230722_SUBC_249'), ('subclass', 'CS20230722_SUBC_238'), ('subclass', 'CS20230722_SUBC_241'), ('subclass', 'CS20230722_SUBC_307'), ('subclass', 'CS20230722_SUBC_259'), ('subclass', 'CS20230722_SUBC_256')]\n"
     ]
    }
   ],
   "source": [
    "type(nodes_to_drop)\n",
    "print(nodes_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "15cebbe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "baseline_marker_path = scratch_dir / 'mouse_markers_230821.json'\n",
    "baseline_precompute_path = scratch_dir / 'precomputed_stats_ABC_revision_230821.h5'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ffd0a4b-dae0-4d60-9ee2-c4307db533ad",
   "metadata": {},
   "source": [
    "Now we can run the mapping again, using the Whole Mouse Brain precomputed stats and marker lookup files, but telling the mapper to ignore these cell type classes that are not present in the isocortex."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "37803e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "additional_files_dir = data_dir / 'additional_files'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b7765be7-7fc5-47d4-8787-745b6be4ce71",
   "metadata": {},
   "outputs": [],
   "source": [
    "baseline_marker_path = additional_files_dir / 'mouse_markers_230821.json'\n",
    "baseline_precompute_path = additional_files_dir / 'precomputed_stats_ABC_revision_230821.h5'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "dd29f037",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/data/additional_files for input\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/scratch to find the test_set.h5ad and save the output\n",
      "=== Running Hierarchical Mapping 1.6.1 \n",
      "\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "No gene_mapper_db provided. Assuming that query genes have already been mapped to the same species/authority as reference genes.\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 798 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/diff_exp/score_utils.py:303: UserWarning: precomputed stats file does not have 'ge1' data\n",
      "  warnings.warn(\"precomputed stats file does not have 'ge1' data\")\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/data/additional_files for input\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/scratch to find the test_set.h5ad and save the output\n",
      "=== Running Hierarchical Mapping 1.6.1 \n",
      "\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "No gene_mapper_db provided. Assuming that query genes have already been mapped to the same species/authority as reference genes.\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 798 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/diff_exp/score_utils.py:303: UserWarning: precomputed stats file does not have 'ge1' data\n",
      "  warnings.warn(\"precomputed stats file does not have 'ge1' data\")\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/data/additional_files for input\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/scratch to find the test_set.h5ad and save the output\n",
      "=== Running Hierarchical Mapping 1.6.1 \n",
      "\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "No gene_mapper_db provided. Assuming that query genes have already been mapped to the same species/authority as reference genes.\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 798 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/diff_exp/score_utils.py:303: UserWarning: precomputed stats file does not have 'ge1' data\n",
      "  warnings.warn(\"precomputed stats file does not have 'ge1' data\")\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/data/additional_files for input\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/scratch to find the test_set.h5ad and save the output\n",
      "=== Running Hierarchical Mapping 1.6.1 \n",
      "\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "No gene_mapper_db provided. Assuming that query genes have already been mapped to the same species/authority as reference genes.\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 798 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/diff_exp/score_utils.py:303: UserWarning: precomputed stats file does not have 'ge1' data\n",
      "  warnings.warn(\"precomputed stats file does not have 'ge1' data\")\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/data/additional_files for input\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/scratch to find the test_set.h5ad and save the output\n",
      "=== Running Hierarchical Mapping 1.6.1 \n",
      "\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "No gene_mapper_db provided. Assuming that query genes have already been mapped to the same species/authority as reference genes.\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 798 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/diff_exp/score_utils.py:303: UserWarning: precomputed stats file does not have 'ge1' data\n",
      "  warnings.warn(\"precomputed stats file does not have 'ge1' data\")\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/data/additional_files for input\n",
      "Using path: /scratch/mfafouti/ABC_atlas_celltyping/scratch to find the test_set.h5ad and save the output\n",
      "=== Running Hierarchical Mapping 1.6.1 \n",
      "\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "No gene_mapper_db provided. Assuming that query genes have already been mapped to the same species/authority as reference genes.\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 798 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n",
      "/usr/local/lib/python3.12/site-packages/cell_type_mapper/diff_exp/score_utils.py:303: UserWarning: precomputed stats file does not have 'ge1' data\n",
      "  warnings.warn(\"precomputed stats file does not have 'ge1' data\")\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n"
     ]
    }
   ],
   "source": [
    "for file in query_file_list:\n",
    "    query_file = file\n",
    "    !apptainer exec --bind {project_path}:{project_path},/tmp:/tmp \\\n",
    "        {project_path}/docker/celltypemapper.sif \\\n",
    "        python {project_path}/scripts/create_files.py \\\n",
    "        --input_dir {additional_files_dir} \\\n",
    "        --scratch_dir {scratch_dir} \\\n",
    "        --nodes_to_drop '{json.dumps(nodes_to_drop)}' \\\n",
    "        --query {query_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea468e29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: create_files.py [-h] --input_dir INPUT_DIR --scratch_dir SCRATCH_DIR\n",
      "                       [--nodes_to_drop NODES_TO_DROP]\n",
      "create_files.py: error: unrecognized arguments: CS20230722_CLAS_27]]\n"
     ]
    }
   ],
   "source": [
    "!apptainer exec --bind {project_path}:{project_path},/tmp:/tmp \\\n",
    "    {project_path}/docker/celltypemapper.sif \\\n",
    "    python {project_path}/scripts/create_files.py --input_dir {additional_files_dir} --scratch_dir {scratch_dir} --nodes_to_drop {json.dumps(nodes_to_drop)} --query="
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
